src/riemannian_diff.jl

```diff
@@ -0,0 +1,232 @@
+
+"""
+    AbstractRiemannianDiffBackend
+
+An abstract type for diff backends. See [`RiemannianONBDiffBackend`](@ref) for
+an example.
+"""
+abstract type AbstractRiemannianDiffBackend end
+
+"""
+    differential(f::Any, c::Curve, t::Real, backend::AbstractDiffBackend = rdifferential_backend())
+
+Compute the Riemannian differential of a curve of type `c` represented by function `f`
+at time `t` using the given backend.
+"""
+differential(::Any, ::Curve, ::Real, ::AbstractRiemannianDiffBackend)
+
+
+"""
+    gradient(f::Any, rf::RealField, p, backend::AbstractRiemannianDiffBackend = rgradient_backend())
+
+Compute the Riemannian gradient of a real field of type `rf` represented by function `f`
+at point `p` using the given backend.
+"""
+gradient(::Any, ::RealField, ::Any, ::AbstractRiemannianDiffBackend)
```

Comment:
Does this also require the manifold? I would've expected that. Ah Maybe not in general for any field, but for a real field (which would be my main application I think) using

`gradient(M,f,p,backend)`

would be my favourite.

Reply:
The manifold is currently stored in `RealField`. Anyway, maybe you are right and I'm trying too hard to make `gradient` compatible with the do-notation.

By the way, do you think `jacobian(M, N, f, p, backend)` looks OK?

Reply:
We could go for the `do` scheme, but even then the gradient could be `gradient(f, M, p)` (though not that nice compared to our other signatures).

I would've expected your `AbstractMaps` for the Jacobian more, but I think your signature looks great.

Reply:
OK, before committing to a design, one last thing. Let's say I want to compute the differential of a curve on the 2-sphere. This would be `differential(Sphere(2), f, p)` and all information about discretization would be stored in `f`, right?

Reply:
Yes, I would say so. For my cases f would not be discretised even, so if it would be, f should have that information.

Reply:
I've made the changes. Now there seems to be no point in having types for maps so I think I'll remove them.